### DISCUSSION TITLE
`Callable[[T|None],TypeGuard[T]]` is incompatible with `Callable[[int|None],TypeGuard[int]]`?

### DISCUSSION THREAD

**Discussion by felebsp** [python/typing#1530] [Q&A]
I have trouble understanding why mypy complains about `Callable[[T|None],TypeGuard[T]]` being incompatible with `Callable[[int|None],TypeGuard[int]]` when `T` is a `TypeVar` in this context. (The error happens in the last line)

```python
import typing as t
from random import random

T = t.TypeVar('T')
def take_while(
    f: t.Callable[[t.Optional[T]], t.TypeGuard[T]], it: t.Iterator[t.Optional[T]]
) -> t.Iterable[T]:
    for x in it:
        if f(x):
            yield x
        else:
            return

def is_not_none(x: t.Optional[T]) -> t.TypeGuard[T]:
    return x is not None

def get_iterator() -> t.Iterable[t.Optional[int]]:
    while True:
        if random() > 0.2:
            yield int(random() * 100)
        else:
            yield None

take_while(is_not_none, iter(get_iterator()))
```
Here's the code in the mypy playground https://mypy-play.net/?mypy=latest&python=3.11&gist=59ff502958f7588377cdba0efd3044db

I'd be thankfull if someone helped me understand `mypy`'s reasoning

**Comment by erictraut** [comment]
This type checks without error [in pyright](https://pyright-play.net/?code=JYWwDg9gTgLgBDAnmYA7A5nAhgZwQKADMoIQ4otUATUuUSWcymkffAFTgF4EA6d5AFMAaligAKAETtJASjZVBhBFgDWggPoB3ABbAANoPH44puIQBcfAMJZ9%2BrACNDAbRcxeAeTAxgEVHYu7AC6wQA0fAJgggDiAK5iVEGhEcAwVh4AkjCCFDDQ7l4%2BfgH6ycH4snAAtAB8fNm5Tq4hFiZmhNBwAB50qHTp7WZmwMqE4t2ybcMzZojAgvpUPUMziziC07MzUIIwcVCoCkp0OBqoEDDn-kbdGUW%2B-oEhVXWRQvGJyVumu-uHPVOcAu8AAcjdjsp0HsNGkmvkJK96lkchRnIJCt5HqUXGgYKEfnBdAZBHB2FA4ptVsNRkxqKRxFV6gAGXgAJkJ2zg80Wyzx4go9JAjLgACo4ABGZnM%2BRc9ZUrnDHlLODg1CCNgwNSaYmGcTAM4g67q1Ko8TQq5wvLQRmyeRAA). I recommend filing a bug in the mypy issue tracker.
